test/functional/p2p_cfilters.py

```diff
@@ -0,0 +1,267 @@
+
+class CFiltersClient(P2PInterface):
+    def __init__(self):
+        super().__init__()
+        # Store the cfilters received.
+        self.cfilters = []
+
+    def pop_cfilters(self):
+        cfilters = self.cfilters
+        self.cfilters = []
+        return cfilters
+
+    def on_cfilter(self, message):
+        """Store cfilters received in a list."""
+        self.cfilters.append(message)
+
+class CompactFiltersTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+        self.extra_args = [
+            ["-blockfilterindex", "-peercfilters"],
+            ["-blockfilterindex"],
+        ]
+
+    def run_test(self):
+        # Node 0 supports COMPACT_FILTERS, node 1 does not.
+        node0 = self.nodes[0].add_p2p_connection(CFiltersClient())
+        node1 = self.nodes[1].add_p2p_connection(CFiltersClient())
+
+        # Nodes 0 & 1 share the same first 999 blocks in the chain.
+        self.nodes[0].generate(999)
+        sync_blocks(self.nodes)
+
+        # Stale blocks by disconnecting nodes 0 & 1, mining, then reconnecting
+        disconnect_nodes(self.nodes[0], 1)
+
+        self.nodes[0].generate(1)
+        wait_until(lambda: self.nodes[0].getblockcount() == 1000)
+        stale_block_hash = self.nodes[0].getblockhash(1000)
+
+        self.nodes[1].generate(1001)
+        wait_until(lambda: self.nodes[1].getblockcount() == 2000)
+
+        # Fetch cfcheckpt on node 0. Since the implementation caches the checkpoints on the active
+        # chain in memory, this checks that the cache is updated correctly upon subsequent queries
+        # after the reorg.
+        request = msg_getcfcheckpt(
+            filter_type=FILTER_TYPE_BASIC,
+            stop_hash=int(stale_block_hash, 16)
+        )
+        node0.send_message(request)
+        node0.sync_with_ping(timeout=5)
+        response = node0.last_message['cfcheckpt']
+        assert_equal(response.filter_type, request.filter_type)
+        assert_equal(response.stop_hash, request.stop_hash)
+        assert_equal(len(response.headers), 1)
+
+        # Reorg node 0 to a new chain
+        connect_nodes(self.nodes[0], 1)
+        sync_blocks(self.nodes)
+
+        main_block_hash = self.nodes[0].getblockhash(1000)
+        assert main_block_hash != stale_block_hash, "node 0 chain did not reorganize"
+
+        default_services = NODE_NETWORK | NODE_WITNESS | NODE_NETWORK_LIMITED
+
+        # Check that nodes have signalled expected services.
+        assert_equal(node0.nServices, default_services | NODE_COMPACT_FILTERS)
+        assert_equal(node1.nServices, default_services)
+
+        # Check that the localservices is as expected.
+        assert_equal(int(self.nodes[0].getnetworkinfo()['localservices'], 16),
+                     default_services | NODE_COMPACT_FILTERS)
+        assert_equal(int(self.nodes[1].getnetworkinfo()['localservices'], 16),
+                     default_services)
```

Comment:
This doesn't seem to be the right place to test default node services... Suggest initialising `default_services` to `node1.nServices`, and checking that it excludes `NODE_COMPACT_FILTERS`.